### DISCUSSION TITLE
Load SWF on the fly.

### DISCUSSION THREAD

**Discussion by balaedit** [airsdk/Adobe-Runtime-Support#2448] [General]
It is possible to play swf file while downloading from server/amazon. 
Means if some part of file is downloaded/buffered like video can it be playing in my app?

Thanks

**Comment by Devin-Licastro** [comment]
I'm not sure about a swf file, but my app uses the feathers video player to stream videos from our AWS servers.
https://feathersui.com/learn/as3-starling/video-player/
You'll probably need to provide more info on your setup and exactly what you want to achieve to get better help.

**Reply by balaedit** [reply]
We are in E-learning Content which had multiple files needs to download. We are first downloading the SWF files from the AWS to local desktop/android and then load it in the movieclip in the main SWF. Just I need to is that while downloading the child swf, I need to play the child swf when some part is downloaded.

**Reply by ajwfrost** [reply]
I would have thought this is standard behaviour if you're using a Loader object, the SWFs were designed so that once the first frame has been received, they can be used...

So it depends a little on what you're doing/how you're doing it, and ensuring that the child SWF is structured so that whatever you need from it is all at the start of the file..

thanks

**Reply by balaedit** [reply]
As with the help of starling's feather we can play the video part of the downloaded video even when the rest of the file is downloading/buffering. Can you @ajwfrost share the sample example of how to play part of downloaded swf even when the rest of the file is downloading. Thanks in advance.

**Reply by ajwfrost** [reply]
From having a quick play with this, we can see the loaded SWF if we add it to the display list. e.g. with:
```
			var loader : Loader = new Loader();
			addChild(loader);
			loader.contentLoaderInfo.addEventListener(Event.COMPLETE, function(e:Event):void { trace("LOADER COMPLETE"); });
			loader.contentLoaderInfo.addEventListener(ProgressEvent.PROGRESS, function(e:ProgressEvent):void { trace("LOADER PROGRESS " + (e.bytesLoaded * 100 / e.bytesTotal) + "%"); });
			var ctx : LoaderContext = new LoaderContext(false, ApplicationDomain.currentDomain);
			ctx.allowCodeImport = true;
			loader.load(new URLRequest("http://localhost:8000/DynamicallyLoaded.swf"), ctx);
			trace("Started loading");
```
We're seeing trace output where the loading SWF completes it's first frame when the swf is only 63% loaded. (The loading process seems to be quite high priority/quick, sometimes it gets to 100% before we see the first frame displayed... but it's only a dummy SWF file so not likely to be as large/complex as yours...)

The key thing for the SWF that you're loading is that the complex stuff shouldn't be exported for frame 1... so if you look at the SWF in a decompiler or SWF investigator, you should see only very brief stuff before the first "show frame" segment. i.e. anything that's available before the "show frame" is stuff you can use (see also the "init" event which will be fired immediately following the processing of the first frame in the loaded file, so within/after that you can then access any class definitions from before the first "show frame").

![image](https://user-images.githubusercontent.com/6400804/216015270-aac3fd55-b95e-4e7c-a62f-034cc3d9a16c.png)

Hope that helps..?

**Reply by balaedit** [reply]
Thanks for it.. Will test on multiple instances.

**Reply by balaedit** [reply]
It worked with my old functionality fine. With updates in my app having encrypted swf failed to do so.
@ajwfrost Loader class work only with swf, jpg, gif and png which is compatible. My swf files are encrypted swf, while downloading I want to change/tweak the header which is nearby between byte 3 to 8. Is it possible with loader class to edit bytearray while downloading the file.

**Reply by ajwfrost** [reply]
Hi

Just to check here .. how are you encrypting the SWF? I think the options would be:
1) if you're downloading your SWFs from somewhere secure and you have authentication (or some way to identify that the client is your app), then they could just be loaded in directly i.e. rely on the HTTPS encryption?
2) if you're using the "native" SWF encryption that we developed - and that has been hacked already - then I would have hoped that the normal `Loader.load()` method works with it..? Is this what you are trying here?
3) alternatively you can download and decrypt a SWF using a different mechanism and then load in via `Loader.loadBytes()`, but this seems to want the whole SWF to be provided at once (although I've not actually verified that and the documentation doesn't say for sure..)

If you're able to clarify the use case, we can try to see what's best here.

thanks

**Reply by balaedit** [reply]
Thanks for reply.

I had mailed you (@ajwfrost on andrew.Frost@harman.com) due to some privacy.

Please check it and guide.

**Reply by ajwfrost** [reply]
Thanks for that.. so I think for this approach, or anything similar, it's going to need the `Loader.loadBytes()` to be used. But this only (currently) works if the SWF is provided as a single, full ByteArray (we just checked that.. the first chunk of the SWF doesn't do much, and subsequent ones cause an error 2124..)

We can take a look to see if this could be adapted to also work with chunked data, so it works more like the normal 'load' or the root swf, where it's processed as it's read. 

Otherwise - the only thing I think of currently is that you may need to just re-architecture into smaller blocks/SWFs and encrypt each of these so that you can load in just what you need to start with...

FYI -> we're about to publish encrypt/decrypt AS3 functions that take a byte array, key and iv, which may mean you can do this a bit more securely e.g. within your host SWF, generate some random bytes and use that as an IV that you can pass to the server, the server can use this to encrypt the SWF and return the encrypted SWF which your host SWF can then decrypt. (There's still the slight security issue where someone could then do a data dump of the process and find the decrypted SWF contents within the application memory... but it's never going to be completely secure from the user who owns the computer it runs on!)

thanks
